plugins/texteditor.koplugin/main.lua

```diff
@@ -179,6 +182,29 @@ Export text to QR code, that can be scanned, for example, by a phone.]]),
                     callback = function()
                         self.qr_code_export = not self.qr_code_export
                     end,
+                    separator = true,
+                },
+                {
+                    text = _("Clean text editor history"),
+                    enabled_func = function()
+                        return #self.history > 0
+                    end,
+                    keep_menu_open = true,
+                    callback = function(touchmenu_instance)
+                        UIManager:show(ConfirmBox:new{
+                            text = _("Clean text editor history?"),
+                            ok_text = _("Clean"),
+                            ok_callback = function()
+                                self.history = {}
+                                self.last_view_pos = {}
+                                -- remove history items from the parent menu
+                                while #touchmenu_instance.item_table_stack[2] > 3 do
+                                    table.remove(touchmenu_instance.item_table_stack[2])
```

Comment:
This feels a bit hacky, hacking TouchMenu internals.
Wouldn't something like what we do to remove one item from history work?
https://github.com/koreader/koreader/blob/2c1cae64b61ba20abd9f7f5d6dac27abdadde35e/plugins/texteditor.koplugin/main.lua#L232-L241
with just looping from the end, just checking for an id, an no break :)
```lua
    for j=#sub_item_table, 1, -1 do 
         if sub_item_table[j]._texteditor_id then 
             table.remove(sub_item_table, j) 
         end 
    end 
    touchmenu_instance:updateItems() 
```

Reply:
Do we have an access to the parent menu sub_item_table from that point?

Reply:
No, but may be this simple change will make the above snippet work:
```diff
--- a/plugins/texteditor.koplugin/main.lua
+++ b/plugins/texteditor.koplugin/main.lua
@@ -97,7 +97,8 @@ end
 function TextEditor:getSubMenuItems()
     self:loadSettings()
     self.whenDoneFunc = nil -- discard reference to previous TouchMenu instance
-    local sub_item_table = {
+    local sub_item_table
+    sub_item_table = {
         {
             text = _("Settings"),
             sub_item_table = {
```
It defines `sub_item_table` as a variable that is now defined _before_ you create its content (the table), so the functions you define inside the table can access this variable.
(Note that on the last line above, the 2nd `sub_item_table = {`, this is the name of a key, not the variable we just defined, so no collision I think.)

Reply:
Thank you, that works.